\chapter{Introduction}

\section{Overview and Aims}

Population genetics is the study of genetic variation within and among populations, and the evolutionary forces shaping that variation. Evolutionary forces such as selection, genetic drift, mutation, and gene flow are studied using allele frequencies and the relationships between alleles across the genome throughout a population. While population genetics is a longstanding field, with origins in the discovery of genetics itself, it has been fundamentally limited by access to computational resources and data needed to better characterize genetics at the population scale. Recently, however, the field has become data-driven in a meaningful way. What used to take weeks for dozens of simulations has evolved into hours for thousands, practitioners use state of the art computational methods, and the field has an abundance of advancements at any given time.

This thesis aims to describe a body of work that contributes both theoretically and practically to population genetics by addressing a series of problems in current population genetics inference tasks. Each section describes a novel data representation and the resulting models that were designed to take advantage of these more performant data representations.

\subsection{The Impact of Population Genetics}

Population genetics provides insights in scenarios that require understanding the changing dynamics of populations of organisms such as agriculture, medicine, and conservation.

Agriculture is perhaps the earliest implementation of population genetics, as genetics was initially described by the works of Mendel on pea plants \cite{mendel_versuche_1866}. This fundamental set of experiments led to the development of Mendelian genetics and the laid the groundwork for genetics as a whole, giving some sort of explanation for the phenomena agriculturalists had noted for many generations prior. Modern crop breeding relies heavily on understanding population genetics for driving optimizations in yield \cite{maccaferri2008quantitative}, disease and pest resistance \cite{poland2009shades, derbyshireBioinformaticDetectionPositive2020, hawkinsEvolutionaryOriginsPesticide2019}, and maintaining diversity in crop populations \cite{hamblin_population_2011}. By understanding the evolutionary forces driving change in a target crop breeders can manipulate crop populations in a way that is beneficial. 

Medical researchers have long been interested in the relationship between ancestry and disease, and the current push for personalized medicine and targeted approaches encourage the use of population-scale analysis for improved medical outcomes \cite{weischenfeldtPhenotypicImpactGenomic2013, ostrowCancerEvolutionAssociated2014}. When the ancestry and population history of a potentially health-affecting locus is identified it creates the opportunity for a more optimized treatment plan taking into account genetic factors previously difficult to ascertain \cite{schuleParkinsonDiseaseAssociated2017}. On the opposite end of the spectrum, population genetics is essential in modeling pathogen trajectories throughout time, which proved essential in understanding the dynamics of pathogenicity in the COVID-19 epidemic where it was used extensively to model the selection of new strains in specific populations and as a whole \cite{keplerDecomposingSourcesSARSCoV22021, wangPervasivePositiveSelection2021}.

Lastly, population genetics is useful as a descriptive tool for understanding the diversity of natural populations and the ways in which they change over time, revealing insights about evolution itself in the process \cite{duttaLossHeterozygositySpectrum2022, pelletierComplexitiesRecapitulatingPolygenic2022, diopBehavioralCostOverdominance2015}. This is particularly true in humans, which have a large amount of literature attempting to identify the genetic history of both individual populations and humans as a species \cite{voightMapRecentPositive2006, fengzhangCopyNumberVariation2009, sverrisdottirDirectEstimatesNatural2014, liInferenceHumanPopulation2011}

\subsection{Population Genetics as a Data-Driven Field}

Population genetics has been fundamentally altered by the development of high throughput sequencing technologies, making it possible to generate incredible amounts of genetic data at a fraction of the cost of previous sequencing methodologies \cite{zhangBiobankscaleInferenceAncestral2023a, kelleherInferringWholegenomeHistories2019, schlottererCombiningExperimentalEvolution2015}. High throughput sequencing has allowed for population scale genetics data, which comes with more accurate approximations of population allele frequency, the foundation for population genetics analyses. Recent decades have seen huge leaps in computational power beneficial to both bioinformatics methods processing genetic data, as well as an influx of population genetics tools to make inferences about the evolutionary history of a large number of samples \cite{kelleherInferringWholegenomeHistories2019, gaoNewSoftwareFast2016, koflerPoPoolation2IdentifyingDifferentiation2011, rosenzweigPowerfulMethodsDetecting2016}. At the same time these advancements in computation have allowed for the recent influx of artificial intelligence applications in genetics, including machine learning (ML) and deep learning (DL) spanning all aspects of analysis from RNA-seq to variant calling \cite{korfmannDeepLearningPopulation2023a}. Machine learning methods are able to link patterns and infer relationships in data, especially highly dimensional and complex data such as genetics \cite{erhanScalableObjectDetection2013}. Population genetics has been utilizing such methods for around a decade at this point \cite{sheehanDeepLearningPopulation2016}, and the amount of machine learning related methods published for population genetics research increases by the day. Machine learning methods are scalable, useful for many different tasks, and generally increase in accuracy with more available training data \cite{lecunDeepLearning2015}. This makes them a highly attractive method for many scientific applications, including population genetics, where they've been used for identifying genetic structure, signatures of positive selection, and even predicting phenotypic outcomes using genotype data \cite{korfmannDeepLearningPopulation2023a}. 

Along with the paradigm shift into machine learning applications have come an influx in data representations driven by the computational burden and cheapness of sequencing and storing genetic data \cite{kelleherEfficientPedigreeRecording2018}. Machine learning methods are highly dependent on how their input data is represented: convolutional neural networks are mostly used for images, while recurrent neural networks are used for serialized data, for example. A network can only be as good as the information provided by the features it is given, therefore it is imperative to explore novel forms of data representation and storage for both computational efficiency and accuracy gains, especially in the area of genetics which relies heavily on flat files storing large amounts of tabular data. 

\subsection{Population Genetics Inference Methods}

Population genetics inference is the fundamental task in population genetics, to infer an evolutionary phenomena by using the observable result of said phenomena: genetic data \cite{sheehanDeepLearningPopulation2016, flagelUnreasonableEffectivenessConvolutional2019}. Since an individual’s or population’s complete history is impossible to thoroughly characterize, inference is a necessary step to further our understanding of how populations change over time. A variety of families of inference methodologies have been developed in the many decades population genetics has existed as a formalized field, these are discussed by general grouping in the following sections. \\

\noindent \textit{Statistical Approaches}

Many population genetics inference methods rely on creating a summary statistic from genetic data \cite{hejaseSummaryStatisticsGene2020}. In its most general form, a statistic simply describes some aspect of the data using a mathematical transformation. Summary statistics are prone to data loss by nature as they are compressing a more whole data representation into a more condensed one, however despite this statistics have been shown to be useful for a large swathe of tasks \cite{arnabUncoveringFootprintsNatural2022, federIdentifyingSignaturesSelection2014, liNewTestDetecting2011, jensenDistinguishingSelectiveSweeps2005, nielsenGenomicScansSelective2005}. Most statistics in population genetics summarize some aspect of the variation in some section of the genome across a sampled number of individuals from one or more populations. This variation can lead to simple descriptive statistics such as $\pi$, a straightforward measure of the average nucleotide diversity in a given region of the genome \cite{neiDNAPolymorphismDetectable1981, neiMathematicalModelStudying1979}, to statistics such as iHS \cite{voightMapRecentPositive2006}, a measure of haplotype block length distributions that is used to detect selective sweeps. Statistics have the benefit of being mathematically defined - any transformation done to achieve the resulting value(s) is completely traceable. Because they are able to be customized, statistics can be developed for a variety of tasks, and can be developed to use a variety of data formats such as single or multi timepoint data as will be discussed in Chapter 1. \\

\noindent \textit{Parameter-Fitting Models and Simulations}

Due to the use of so many summary statistics, a common approach to population genetics inference is the use of frameworks that simulate large amounts of data under a variety of parameterizations and identify those which most closely match the target empirical data. Population genetic simulations are a core component of the researcher’s toolbox from the early stages of the field \cite{tajimaStatisticalMethodTesting1989}, and allows one to test hypotheses at scale under a known set of assumptions. Population genetic simulations come in two forms: forward in time, which explicitly models some number of individuals’ genetics and the evolutionary forces acting on them generation after generation \cite{hallerSLiMForwardGenetic2019}, and backward in time, known as coalesent simulations \cite{kernDiscoalFlexibleCoalescent2016, kelleherEfficientCoalescentSimulation2016, excoffierFastsimcoal2DemographicInference2021}. While forward in time simulations are valuable for their explicit modeling of individuals and populations, coalescent simulations are extremely fast due to the underlying statistical assumptions allowing for backwards in time simulation, leading to many simulation-heavy experimental frameworks overwhelmingly using coalescent simulations simply due to computational limitations \cite{marthAlleleFrequencySpectrum2004, liInferenceHumanPopulation2011}.

Specific methodologies of population genetic inference using simulations and summary statistics can include multiple different frameworks, one of the most notable being Approximate Bayesian Computation (ABC) \cite{beaumontApproximateBayesianComputation2002}. In brief, ABC works by identifying a value or distribution of values of a specific set of summary statistics drawn from the target empirical data we wish to infer about. A demographic model is created and simulated under a swathe of parameterizations such as population size, selection coefficient, etc, and the same set of summary statistics are drawn from the resulting simulations. Simulations that result in summary statistics more closely resembling the target empirical data are assumed to more closely resemble the actual demographic model of the target data, and the parameters used to get them establish a distribution of likely values. From this method a set of parameters that potentially accurately describe the demographic history of the target population can be found. While ABC is highly effective in many cases , both in its original form \cite{beaumontApproximateBayesianComputation2002} and the variety of modifications developed for the method \cite{follWFABCWrightFisher2015, pudloReliableABCModel2016, ferrer-admetllaApproximateMarkovModel2016, raynalABCRandomForests2019}, it remains an extremely computationally intensive task. Methods such as ABC (as well as contemporary methods such as hidden Markov models (HMMs)) suffer from long and intensive runtimes as well as a problem known as the “curse of dimensionality” . This concept describes a tradeoff in some methods between the dimensionality of the data and the accuracy of inference. For example, when modeling highly dynamic populations with many varying parameters over time, ABC will have to optimize over such a large search space that it may never fully converge on a reasonable distribution of parameters useful for examination \cite{beaumontApproximateBayesianComputation2002}. As demographic models and inference tasks grow in complexity, alternative methods that can overcome this limitation have been developed, such as machine learning.


\subsection{Existing Machine Learning Applications in population Genetics}

Machine learning in population genetics has become an area of interest over the past decade, driving advances in the field using a variety of models and approaches. Once such approach called supervised machine learning (SML) uses labeled training data to learn relationships and has become a common approach \cite{korfmannDeepLearningPopulation2023a}. At its core SML in population genetics works similarly to previous parameter estimation methods: it takes in some characteristic of the target population, generates simulations to see how an array of parameterizations affect said simulations, and infers the closeness of fit between the resulting simulations and the target empirical data \cite{schriderSupervisedMachineLearning2018}. SML, much like ABC, relies heavily on simulated data. However, due to the nature of machine learning architectures, particularly deep learning approaches, these methods learn patterns directly from the raw data rather than the myriad of summary statistics and assumptions made by related approaches \cite{lecunBackpropagationAppliedHandwritten1989, lecunDeepLearning2015}. The general workflow for a given population genetics inference method involves simulating data to best match the known demographic history of the target population, formatting and training a model using said data, and predicting using the real data \cite{schriderInferringSelectiveConstraint2015, rayIntroUNETIdentifyingIntrogressed2023, sheehanDeepLearningPopulation2016, schriderHICRobustIdentification2016, wangAutomaticInferenceDemographic2021, kernDiploSHICUpdated2018, moDomainadaptiveNeuralNetworks2023}. This approach has shown to be highly effective in a variety of tasks and is even suitable for using directly on genotype matrices, the simplest form of population genetic data representation \cite{flagelUnreasonableEffectivenessConvolutional2019}.

Initial work with NNs proved that it was a powerful class of tools for population genetics \cite{sheehanDeepLearningPopulation2016}. The “expressiveness” of NNs (a models ability to understand high-dimensional and/or complex data relationships) proves to be highly effective in detecting selection using only summary statistics, and launched a new wave of inference methods adopting similar frameworks. Many early NN implementations used summary statistics as their inputs, which is what the model learned to relate to some target output variable such as the presence or absence of selection \cite{schriderHICRobustIdentification2016}. The next notable advancement in the field came with the introduction of convolutional neural networks (CNNs) \cite{lecunBackpropagationAppliedHandwritten1989} applied directly to genotype matrices, a binary representation of the genetic variants in a given region of the genome across some number of sampled individuals \cite{flagelUnreasonableEffectivenessConvolutional2019}. Without the need for summary statistics a given model was able to implicitly learn any necessary transformations of the input features to draw conclusions about the optimization task, which reduced the burden of computing a large amount of summary statistics and increased the expressiveness and generalizability of these models for a wide variety of inference tasks . Since then there have been many uses of models learning directly from genotype matrices, however new data representations have begun cropping up both for specific tasks and as more general population genetic data representations \cite{GenomeWideInferenceAncestral, kelleherEfficientCoalescentSimulation2016}. These new representations are indicative of a new potential for developing methods to achieve better accuracy, more efficient computation, and overall better tooling for population genetic inference.

\section{Motivation}

Despite ML in population genetics being a well-studied area there are still many accuracy gains to be made for different inference tasks. As data sources and questions constantly evolve, so too must the methodologies to better understand that data. The methods described in this thesis aim to provide novel approaches for multiple population genetics and adjacent tasks, filling in knowledge gaps by utilizing novel data representations and state of the art neural network approaches. These approaches aim to not only provide accuracy increases for their target tasks, but to do so in a computationally efficient manner that democratizes and encourages their use in as many contexts as possible.

\section{Structure}

This thesis describes three novel methods that utilize ML to further the understanding of genomics and population genetics. First  it describes a method for time-series sequencing analysis, allowing for more accurate detection of positive selective sweeps using a paradigm of multiple sequencing runs over a time course of a population. Next, it describes a novel approach for using graph convolutional networks applied to tree sequences and how this approach compares to the current state of the art convolutional neural network approaches. And finally it discusses a new approach for detecting and genotyping structural variation and transposable elements using short read data for population-scale structural variant analysis.
