%The word �Abstract� should be centered 2? below the top of the page. 
%Skip one line, then center your name followed by the title of the 
%thesis/dissertation. Use as many lines as necessary. Centered below the 
%title include the phrase, in parentheses, �(Under the direction of  
%_________)� and include the name(s) of the dissertation advisor(s).
%Skip one line and begin the content of the abstract. It should be 
%double-spaced and conform to margin guidelines. An abstract should not 
%exceed 150 words for a thesis and 350 words for a dissertation. The 
%latter is a requirement of both the Graduate School and UMI's 
%Dissertation Abstracts International.
%Because your dissertation abstract will be published, please prepare and 
%proofread it carefully. Print all symbols and foreign words clearly and 
%accurately to avoid errors or delays. Make sure that the title given at 
%the top of the abstract has the same wording as the title shown on your 
%title page. Avoid mathematical formulas, diagrams, and other 
%illustrative materials, and only offer the briefest possible description 
%of your thesis/dissertation and a concise summary of its conclusions. Do 
%not include lengthy explanations and opinions.
%The abstract should bear the lower case Roman number ii (if you did not 
%include a copyright page) or iii (if you include a copyright page).

\begin{center}
\vspace*{52pt}
{\normalfont\textbf{ABSTRACT}}
\vspace{11pt}

\begin{singlespace}
Logan Scott Whitehouse: Deep Learning Applications in Population Genetics \\
(Under the direction of Daniel Schrider)
\end{singlespace}
\end{center}
In recent years population geneticists have finally started to obtain the amount of data needed to test decades-old hypotheses. With the simultaneous advent of machine learning in genetics it is now more critical than ever to consider the impact of machine learning methods in the space, and how we can best use these methods to aid in the goal of better understanding evolution through the lens of genetics. One common factor in both machine learning and population genetics is the myriad of choice when it comes to data representation; multiple sequence alignments, tree sequences, site frequency spectrums, all differing representations of the same information packaged in different ways. This thesis demonstrates the impact of feature engineering for population genetics data and multiple approaches for developing machine learning models to best utilize these representations. First discussing a data representation of time-series genomics data, a model is built for detecting selection when a population can be sampled throughout time to better understand the temporal aspects of positive selection. Then a model is discussed that can be used in a variety of population genetics applications, taking advantage of computationally efficient tree sequences and their inherent graph structure to perform population genetics inference. Finally a set of approaches for characterizing and detecting structural variants using short-read sequencing data with the aim of population-scale structural variant analysis is discussed. The insights gained by these experiments and the methods developed therein will add to the landscape of useful population genetics tools and provide a stepping stone for further method development moving forward. 
\clearpage
