%The word �Abstract� should be centered 2? below the top of the page. 
%Skip one line, then center your name followed by the title of the 
%thesis/dissertation. Use as many lines as necessary. Centered below the 
%title include the phrase, in parentheses, �(Under the direction of  
%_________)� and include the name(s) of the dissertation advisor(s).
%Skip one line and begin the content of the abstract. It should be 
%double-spaced and conform to margin guidelines. An abstract should not 
%exceed 150 words for a thesis and 350 words for a dissertation. The 
%latter is a requirement of both the Graduate School and UMI's 
%Dissertation Abstracts International.
%Because your dissertation abstract will be published, please prepare and 
%proofread it carefully. Print all symbols and foreign words clearly and 
%accurately to avoid errors or delays. Make sure that the title given at 
%the top of the abstract has the same wording as the title shown on your 
%title page. Avoid mathematical formulas, diagrams, and other 
%illustrative materials, and only offer the briefest possible description 
%of your thesis/dissertation and a concise summary of its conclusions. Do 
%not include lengthy explanations and opinions.
%The abstract should bear the lower case Roman number ii (if you did not 
%include a copyright page) or iii (if you include a copyright page).

\begin{center}
\vspace*{52pt}
{\normalfont\textbf{ABSTRACT}}
\vspace{11pt}

\begin{singlespace}
Logan Scott Whitehouse: Deep Learning Applications in Population Genetics \\
(Under the direction of Daniel Schrider)
\end{singlespace}
\end{center}
In recent years population geneticists have finally started to obtain the amount of data needed to test decades-old hypotheses. With the adoption of machine learning applications to problems in genetics it is now more critical than ever to consider the impact of such methods in this space, and how we can best use them to aid in the goal of better understanding patterns and processes in evolutionary genetics. One common factor in both machine learning and population genetics is the amount of choice when it comes to data representation: multiple sequence alignments, tree sequences, site frequency spectrums, are all differing representations of the same information packaged in different ways, each of which coming with tradeoffs for utility, feature selection, and their effect on downstream inference. This thesis demonstrates the impact of feature engineering for population genetics data and multiple approaches for developing machine learning models to best utilize these representations. First we discuss a data representation of time-series genomics data and accompanying model built for detecting selection used when a population can be sampled throughout time to better understand the temporal aspects of positive selection. We then discuss graph convolutional networks in population genetics inference by taking advantage of computationally efficient tree sequences and their inherent graph structure, and our multiple benchmarks against the more commonly-used convolutional neural network architecture and multiple sequence alignment. Finally we discuss a set of approaches for characterizing and detecting structural variants and transposable elements using short-read sequencing data, where we aim to develop a method for population-scale structural variant analysis by extracting large amounts of information from alignments to better inform variant discovery and characterization. The insights gained by these experiments and the methods developed therein will add to the landscape of useful population genetics tools and provide a stepping stone for further method development moving forward. 
\clearpage
